101-vm-sshkey/azuredeploy.json

```diff
@@ -2,47 +2,62 @@
   "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json",
   "contentVersion": "1.0.0.0",
   "parameters": {
+    "projectName": {
+      "type": "string",
+      "metadata": {
+        "description": "Specifies a name for generating resource names."
+      }
+    },
+    "location": {
+      "type": "string",
+      "defaultValue": "[resourceGroup().location]",
+      "metadata": {
+        "description": "Specifies the location for all resources."
+      }
+    },
     "adminUsername": {
       "type": "string",
       "defaultValue": "azureuser",
       "metadata": {
-        "description": "User name for the Virtual Machine."
+        "description": "Specifies a username for the Virtual Machine."
       }
     },
-    "sshKeyData": {
+    "adminPublicKey": {
       "type": "string",
       "metadata": {
-        "description": "SSH rsa public key file as a string."
+        "description": "Specifies the SSH rsa public key file as a string. Use \"ssh-keygen -t rsa -b 2048\" to generate your SSH key pairs."
       }
     }
   },
   "variables": {
-    "vmSize": "Standard_DS2_v2",
-    "vmName": "sshvm",
-    "ubuntuOSVersion": "16.04-LTS",
-    "uniqueDnsLabelPrefix": "[concat('store', uniquestring(resourceGroup().id))]",
-    "location": "[resourceGroup().location]",
-    "imagePublisher": "Canonical",
-    "imageOffer": "UbuntuServer",
-    "addressPrefix": "10.0.0.0/16",
-    "subnet1Name": "Subnet-1",
-    "subnet1Prefix": "10.0.0.0/24",
-    "nicName": "sshNIC",
-    "publicIPAddressName": "sshPublicIP",
-    "publicIPAddressType": "Dynamic",
-    "virtualNetworkName": "sshVNET",
-    "networkSecurityGroupName": "networkSecurityGroup1",
-    "sshKeyPath": "[concat('/home/',parameters('adminUsername'),'/.ssh/authorized_keys')]",
-    "vnetID": "[resourceId('Microsoft.Network/virtualNetworks', variables('virtualNetworkName'))]",
-    "subnet1Ref": "[concat(variables('vnetID'),'/subnets/',variables('subnet1Name'))]",
-    "apiVersion": "2015-06-15"
+    "networkInterface": {
```

Comment:
let's leave the variables flat - the object variables add some complexity and tooling does not handle these well today.

There are a number of single use variables which we can just remove and replace with literals (like you did for vm image info)

Reply:
fixed. I used literals for the publicIPAddresses sku. Let me know it needs to be a parameter.